### DISCUSSION TITLE
Attiny84 Serial change pins

### DISCUSSION THREAD

**Discussion by Flo082002** [SpenceKonde/ATTinyCore#642] [Support / Q & A]
Hi,
is it possible to change the pins of the serial interface on Attiny84?
I don't want to use the SoftwareSerial library because it uses much more RAM and ROM than the standard Serial of the attiny.

**Comment by SpenceKonde** [comment]
~You can change TX - call Serial.setTxBit(bit number). Bit number can be 0, 1 (default), 3, 4, 5, 6, or 7~

~This is the number of the pin within port. That is you can use PA0, PA1, PA3, PA4, PA5, PA6, or PA7. PA2 is RX, and expanding it to use a different port would entail additional logic, hence increase overhead.~
You should be able to do that.... but there's a typo in that function that nobody caught.

IT will be fixed for 2.0.0

RX cannot be moved. There are three ways to get an interrupt from a pin changing state. 
1. PCINT on any pin. That's a huge loser for a software serial port, because then you can't do any pinchange interrupts yourself, because you would get duplicate vector errors like you do with software serial. So that's a non-starter. For one thing, it would break most common sleep mode use cases..... 
2. INT0 on the INT0 pin (some parts have INT1 or INT2, but not that one). These full featured external interrupt pins are the only one(s) that can be used with attachInterrupt. So we don't want to block that off. 
3. Finally, there is the trick we use: Configuring the Analog Comparator to compare AIN1 with the internal bandgap reference. and firing the interrupt on that. The AC is almost never used on classic AVRs, and so it is very unlikely to interfere with user code, and it's usually on a pin that isn;t very useful for other things (for example, on the tiny84, the only special functionality of the pins used for the software serial port are two analog inputs (there are 6 others), and the two analog comparator pins, which cannot be used for anything else anyway when the builtin softwareserial implementation is in use. (obviously, it makes me wonder why you want to move them off of two of the least useful pins on the chip!) . 

The choice of AIN0 (PA1 on '84) for TX is arbitrary, but it's always adjacent to AIN1, which must be RX, and it's something that I can write in the documentation "the builtin software serial port is on AIN0 (tx) and AIN1 (rx)", and then even if they aren't looking at my versions of the pinout charts, it doesn't matter, because it;'s shown on the datasheet pinout charts too. 

2.0.0 will introduce a menu option to turn off RX functionality from a tools menu option, which will have the benefit of allowing TX to be used, without any flash overhead with an unused RX mechanism. since many users are only doing TX (often for debug) with it),  including allowing the interrupt (and hence the analog comparator in general) to be usesd for some other purpose..
